zrml/prediction-markets/src/migrations.rs

```diff
@@ -16,11 +16,141 @@
 // You should have received a copy of the GNU General Public License
 // along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
 
+use crate::{
+    Config, MarketIdsPerOpenBlock, MarketIdsPerOpenTimeFrame, MarketsCollectingSubsidy,
+    Pallet as PredictionMarkets,
+};
+use core::marker::PhantomData;
+use frame_support::{
+    log,
+    pallet_prelude::{StorageVersion, Weight},
+    traits::{Get, OnRuntimeUpgrade},
+};
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "try-runtime")] {
+        use alloc::vec::Vec;
+    }
+}
+
+const PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION: u16 = 7;
+const PREDICTION_MARKETS_NEXT_STORAGE_VERSION: u16 = 8;
+
+pub struct DrainDeprecatedStorage<T>(PhantomData<T>);
+
+impl<T> OnRuntimeUpgrade for DrainDeprecatedStorage<T>
+where
+    T: Config,
+{
+    fn on_runtime_upgrade() -> Weight {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let prediction_markets_version = StorageVersion::get::<PredictionMarkets<T>>();
+        if prediction_markets_version != PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION {
+            log::info!(
+                "DrainDeprecatedStorage: prediction-markets version is {:?}, but {:?} is required",
+                prediction_markets_version,
+                PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION,
+            );
+            return total_weight;
+        }
+        log::info!("DrainDeprecatedStorage: Starting...");
+        let mut reads_writes = 1u64; // For killing MarketsCollectingSubsidy
+        reads_writes =
+            reads_writes.saturating_add(MarketIdsPerOpenBlock::<T>::drain().count() as u64);
+        reads_writes =
+            reads_writes.saturating_add(MarketIdsPerOpenTimeFrame::<T>::drain().count() as u64);
```

Comment:
Since market number is still relatively low, I assume this should be fine. Otherwise [clear_storage_prefix](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/migration/fn.clear_storage_prefix.html) should be used in a multi-block migration.

Reply:
The heaviest migration should be the one in market-commons, which, on Battery Station with 700 markets, should require about 1,400 DB interactions. That amounts to 0.035 seconds, assuming 25 microseconds for each interaction.

So... are multi-block migrations a thing now? If I had known that, I would have written all of these using multi-block tools.

Reply:
battery station: 2023-12-22 21:53:49.374  INFO main try-runtime::cli: TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (136325000000 ps, 0 byte), total weight = (500000000000 ps, 5242880 byte) (27.27 %, 0.00 %).    

mainnet: 2023-12-22 22:22:56.552  INFO main try-runtime::cli: TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (72950000000 ps, 0 byte), total weight = (500000000000 ps, 5242880 byte) (14.59 %, 0.00 %).